### DISCUSSION TITLE
Scope of a top-level let declaration

### DISCUSSION THREAD

**Discussion by seh** [cuelang/cue#1041] [General] [answered]
By my reading of the CUE specification, a [_let declaration_](https://tip.cuelang.org/docs/references/spec/#let-declarations) occurring at the top level of a file in a package should be scoped to the package. The ["Declarations and scope" section](https://tip.cuelang.org/docs/references/spec/#declarations-and-scope) of the specification says that
> 2. The scope of an identifier denoting a field declared at top level (outside any struct literal) is the package block.

It clarifies that aliases occurring at the top level are scoped to the containing file, but it doesn't mention a similar restriction for _let declarations_.

In my experiments with CUE version 0.4.0, though, I find that a _let declaration_ is only visible within the containing file, and not in another file within the same package. If instead I replace the _let declaration_ with a top-level field, I can refer to it from other files within the same package.

Is my interpretation of the specification wrong? Is the CUE tool enforcing the wrong scope?

**Comment by seh** [comment, marked as answer]
Based on experience with similar constructs in other languages, I find the CUE tool's file-scoped enforcement to make more sense, because _let declarations_ seem like _lexical_ elements to me, intended to be confined as locally as possible, to only where referring expressions can "see" the declaration. Should we tighten the specification's language accordingly?

**Reply by myitcv** [reply]
I agree with your analysis. I suspect this is a simple omission from the spec and that point 3 in the "Declarations and Scope" section should read:

> The scope of an identifier denoting an alias or let declaration declared at top level (outside any struct literal) is the file block.

Do you want to create a CL/PR to make this change? @mpvl can weigh in at that point, because this point needs to be clarified in the spec at least.

**Reply by seh** [reply]
> Do you want to create a CL/PR to make this change?

Does the spec document live in the repository that sits behind Gerrit? I found some stub files that appear to build it in [the _cuelang/cuelang.org_ repository](https://github.com/cuelang/cuelang.org), but not the document itself.

_[Time passes ....]_

I found the file _spec.md_ [here](https://github.com/cuelang/cue/blob/master/doc/ref/spec.md). Is that the one to edit?

**Reply by myitcv** [reply]
Yes, you found it!

**Reply by seh** [reply]
Please see [my patch in Gerrit](https://cue-review.googlesource.com/c/cue/+/9981).
(What a twisty path to finally be able to upload a patch successfully!)

**Reply by myitcv** [reply]
Thank you. 

> What a twisty path to finally be able to upload a patch successfully!

Yes, submitting a CL is initially not as obvious/straightforward as a PR, but with practice it's a very effective and powerful workflow, and the other benefits we get from Gerrit (far superior code review, commit message review, support for stacked changes...) make it a no-brainer. We also plan to do a live stream/recording of how to submit a change via Gerrit to support the docs.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/1041.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
